mavros_extras/src/plugins/odom.cpp

```diff
@@ -63,96 +66,103 @@ class OdometryPlugin : public plugin::PluginBase {
-				odom->child_frame_id, "local_origin_ned", ros::Time(0)));
+							odom->child_frame_id, "local_origin" + desired_frame, ros::Time(0)));
+			// transform lookup WRT body frame
 			tf_parent2ned = tf2::transformToEigen(m_uas->tf2_buffer.lookupTransform(
-				odom->header.frame_id, "local_origin_ned", ros::Time(0)));
+							odom->header.frame_id, "local_origin" + desired_frame, ros::Time(0)));
 		} catch (tf2::TransformException &ex) {
-			ROS_WARN("%s",ex.what());
+			ROS_WARN("odom: %s",ex.what());
 			ros::Duration(1.0).sleep();
 			return;
 		}
 
-		// unpack data into Eigen
-		Eigen::Vector3d pos_parent(
-			odom->pose.pose.position.x,
-			odom->pose.pose.position.y,
-			odom->pose.pose.position.z);
-
-		Eigen::Vector3d lin_vel_child(
-			odom->twist.twist.linear.x,
-			odom->twist.twist.linear.y,
-			odom->twist.twist.linear.z);
-
-		Eigen::Vector3d ang_vel_child(
-			odom->twist.twist.angular.x,
-			odom->twist.twist.angular.y,
-			odom->twist.twist.angular.z);
-
-		Eigen::Quaterniond q_parent2child(
-			odom->pose.pose.orientation.w,
-			odom->pose.pose.orientation.x,
-			odom->pose.pose.orientation.y,
-			odom->pose.pose.orientation.z);
-
-		Eigen::Vector3d lin_vel_NED;
-		Eigen::Vector3d ang_vel_ned;
-		Eigen::Vector3d pos_NED;
-		Eigen::Quaterniond q_parent2ned;
-
-		Eigen::Matrix<double, 6, 6> pose_cov_parent;
-		Eigen::Matrix<double, 9, 9> pose_cov_ned;
-		Eigen::Matrix<double, 6, 6> twist_cov_child;
-		Eigen::Matrix<double, 3, 3> att_cov_ned;
-
-		for (int i=0; i<6; i++) for (int j=0; j<6; j++) {
-			pose_cov_parent(i, j) = odom->pose.covariance[i*6 + j];
-			twist_cov_child(i, j) = odom->pose.covariance[i*6 + j];
-		}
+		/** Build 9x9 covariance matrix to be transformed and sent
+		 */
+		ftf::Covariance9d cov_full {};	// zero initialized
+		ftf::EigenMapCovariance9d cov_full_map(cov_full.data());
+
+		cov_full_map.block<6, 6>(0, 0) = ftf::EigenMapConstCovariance6d(odom->pose.covariance.data());
+		cov_full_map.block<3, 3>(6, 6) = Eigen::Matrix3d::Identity();
+
+		/** Build 6x6 covariance matrix to be transformed and sent
+		 */
+		ftf::Covariance6d cov_vel {};	// zero initialized
+		ftf::EigenMapCovariance6d cov_vel_map(cov_vel.data());
+
+		/** Apply linear transforms
+		 */
+		Eigen::Vector3d pos_ned(tf_parent2ned.linear() * ftf::to_eigen(odom->pose.pose.position));
+		Eigen::Vector3d lin_vel_ned(tf_child2ned.linear() * ftf::to_eigen(odom->twist.twist.linear));
+		Eigen::Vector3d ang_vel_ned(tf_child2ned.linear() * ftf::to_eigen(odom->twist.twist.angular));
+		Eigen::Quaterniond q_parent2ned(tf_parent2ned.linear() * ftf::to_eigen(odom->pose.pose.orientation));
+
+		/** Apply covariance transforms */
+		/** Pose+Accel 9-D Covariance matrix
+		 *  WRT world frame (half upper right triangular)
+		 */
+		Eigen::Matrix<double, 9, 9> r_9d;
+		r_9d.block<3, 3>(0, 0) = r_9d.block<3, 3>(3, 3) = tf_parent2ned.linear();
+		r_9d.block<3, 3>(6, 6) = Eigen::Matrix3d::Identity();
```

Comment:
This is only rotating the rotation, but you need to also rotate the velocity as stated in the message, the velocity is given in the child frame.

Also, I think we can use the 9x6 matrix that I setup since we don't use accels.

Reply:
> This is only rotating the rotation, but you need to also rotate the velocity as stated in the message, the velocity is given in the child frame.

Check the code bellow: https://github.com/mavlink/mavros/pull/823/files#diff-750c83d1bf274515183c544f142bac77R132